### ISSUE TITLE
Split ColPixel

### ISSUE THREAD

**Issue by schmie** [schmie/hemeMove#5]
**Reported by hywel on 19 Sep 2011 15:52 UTC**
Currently, we have a single !ColPixel object which holds the drawn pixel data for each component of the vis pipeline (glyper, ray tracer, streaker). This object also acts as the sole object transferred between cores to compose the image. The economy here is that the screen co-ordinates of each pixel are only in memory once per core.

The problems are:
- our communications include unused bytes if we are not using one of the vis components. Ditto for pixels that don't have all 3 kinds of data.
- we may want the data written to and communicated by each component to be dependent on the type of the component (e.g. we have already wanted to template the ray tracer on the way pixels from different cores are to be combined). This requires templating the !ColPixel on the specific component type, which in turn requires templating other components on that specific components type.
- it makes it harder for us to add and remove visualisation components in real time.

I'd like to remove the !ColPixel as a concept, and have each component use its own pixel type for storage and communication. These will only be composed into a single object just before being saved to file or sent to the client. 

Assigning to Rupert for design review.

**Comment by schmie** [comment]
**Comment by rupert on 22 Sep 2011 10:01 UTC**
This is reasonable. I assume this will be a standard pattern of base class with (initially) 4 concrete subclasses, or minor variation thereof.

Assign back to Hywel to implement

**Comment by schmie** [comment]
**Comment by hywel on 22 Sep 2011 17:31 UTC**
Maybe. I'm not yet sure how much commonality there is between the 3 implementations that will initially be required, so there might not be a base class. The fourth implementation you mention is for Nick's work on the ray tracer, which is not yet in the main codebase.

**Comment by schmie** [comment]
**Comment by nick on 23 Sep 2011 12:04 UTC**
> The problems are:
> - our communications include unused bytes if we are not using one of the vis components. Ditto 
> for pixels that don't have all 3 kinds of data.
- The glyphs and streaklines are kept in the most significant bit of the x and y co-ordinates and are therefore free.
- The case of a glyph/streakline without ray data should occur very rarely, because these overlay the ray data. In the diffTest, 34,523 pixels were sent without ray data (I wonder why this happens at all), vs 21,040,794 without. In fact, the best option would probably be not to send these pixels at all, as the streaklines and glyphs should be overlayed on top of ray tracing data.
> - we may want the data written to and communicated by each component to be dependent on the 
> type of the component (e.g. we have already wanted to template the ray tracer on the way 
> pixels from different cores are to be combined). This requires templating the ColPixel on the > specific component type, which in turn requires templating other components on that specific 
> components type.
This will not be alleviated by splitting ColPixel. Whatever structure is sent over the network will need to be templated, as will whatever class is responsible for that. The only way to remove this would be run-time polymorphism. 
> it makes it harder for us to add and remove visualisation components in real time.
Steaklines and glyphs being free, this isn't a problem. If you want to turn off ray tracing and sill have glyphs and streaklines, then yes I'd agree that's a problem.

I do however believe that it's possible that, if non-blocking communication after each component is called, you might get a performance boost as data transfer can begin sooner than otherwise. If this is the intention, I'd agree it's a good idea worth trying out.

If I was doing that, I'd definitely go with Rupert's design for the data as sent between cores, the common components being the x and y co-ordinates, and the ability to merge. ColPixel is also currently responsible for rendering a given pixel, which is something that needs to be considered in the new design. Perhaps you could have a separate interface for that, IRenderable, implemented by each of the pixel types, which allows them to (optionally) take in an existing pixel colour and output a new pixel colour?

**Comment by schmie** [comment]
**Comment by hywel on 23 Sep 2011 13:01 UTC**
That bitfield limits us to a certain size of image. I can imagine situations where we'd want to go beyond that. Also, the streakline data includes 2 floats and an int. The current implementation reads, to me, like a computer scientist fiddling with his bits, if you'll forgive the expression. If you haven't read it before, I highly recommend a Knuth paper called "Literate Programming" which certainly convinces me that the new design is better than the old.

If we're only displaying streaklines on the client, why should we render, send and compose this other stuff at all? I mean this rhetorically - obviously we shouldn't be sending it. Remember too that the diff-test is not the use-case we're aiming for, it's a regression test for the research (i.e. non-steered) case.

Nick, it's possible to have only the class that creates and consumes the data know about the type and nothing intermediate. The templating can therefore end at the boundary to the class. Run-type polymorphism is one way to handle this, encapsulation of the datatype is another (and the one I prefer).

Also note that our designs should not limit themselves to consider rays, glyphs and streaklines. For instance consider how much copy-and-pasting would be involved to have both ray-tracers running in parallel, visualising the results on the steering client (if the answer is > 0 copy-and-pasting then we're doing something wrong). Adding a new imaging technique should be as easy as writing a self-contained class, registering it in one place then running the simulation.

We haven't investigated parallel non-blocking communications. It's possible it would make the code faster, it's also possible that it wouldn't (depending on the MPI implementation - each communication has a constant latency time of about 10 micro s; if it takes 9 micro s to do computation before the next parallel non-blocking communication, we've saved nothing). The central aim of this work is to improve the engineering of the visualisation code. This 

The original design is Marco's, rather than Rupert's. Something similar might be used for the actual communication, but there would be a conversion step beforehand (this is what I was talking about on comment:6:ticket:7). These would be converted back on the other side before being merged. Each pixel would be responsible for merging itself.

You're right that the !ColPixels contain the code for rendering themselves (as well as serializing themselves) and that I hadn't described that here. My intention was to have each pixeltype derive from a sub-class (as Rupert and you have both pointed out, they have functionality in common relating to sending, receiving, merging, and colouring), which would have some function like !GetColour(char[3]). Code in the level above can decide how to combine and serialise the colours.

**Comment by schmie** [comment]
**Comment by nick on 23 Sep 2011 13:39 UTC**
Replying to [hywel](comment:4):
> That bitfield limits us to a certain size of image. I can imagine situations where we'd want to go beyond that. Also, the streakline data includes 2 floats and an int. The current implementation reads, to me, like a computer scientist fiddling with his bits, if you'll forgive the expression. If you haven't read it before, I highly recommend a Knuth paper called "Literate Programming" which certainly convinces me that the new design is better than the old.
Fair point - I'd forgotten about the extra fields for the streakline data, which probably does make the changes much more advantageous. I'm very much in favour of coding to make things look nicer. I had just assumed that our changes weren't allowed to reduce the performance of hemeLB without good reason.

> If we're only displaying streaklines on the client, why should we render, send and compose this other stuff at all? I mean this rhetorically - obviously we shouldn't be sending it. Remember too that the diff-test is not the use-case we're aiming for, it's a regression test for the research (i.e. non-steered) case.
I'd assumed that the streaklines would not be rendered without the ray tracing data, as it'd be hard to infer anything without the vessels in the background.

> Nick, it's possible to have only the class that creates and consumes the data know about the type and nothing intermediate. The templating can therefore end at the boundary to the class. Run-type polymorphism is one way to handle this, encapsulation of the datatype is another (and the one I prefer).
How the latter would work?

> Also note that our designs should not limit themselves to consider rays, glyphs and streaklines. For instance consider how much copy-and-pasting would be involved to have both ray-tracers running in parallel, visualising the results on the steering client (if the answer is > 0 copy-and-pasting then we're doing something wrong). Adding a new imaging technique should be as easy as writing a self-contained class, registering it in one place then running the simulation.
Except that the data into the class varies depending on the type of imaging.
 
> We haven't investigated parallel non-blocking communications. It's possible it would make the code faster, it's also possible that it wouldn't (depending on the MPI implementation - each communication has a constant latency time of about 10 micro s; if it takes 9 micro s to do computation before the next parallel non-blocking communication, we've saved nothing). The central aim of this work is to improve the engineering of the visualisation code. 
Fair enough

> The original design is Marco's, rather than Rupert's. Something similar might be used for the actual communication, but there would be a conversion step beforehand (this is what I was talking about on comment:6:ticket:7). These would be converted back on the other side before being merged. Each pixel would be responsible for merging itself.
Fair enough - this is not far off what's already happening with the ray data, except that the raydata from all the rays are prematurely added to ColPixel.

> You're right that the !ColPixels contain the code for rendering themselves (as well as serializing themselves) and that I hadn't described that here. My intention was to have each pixeltype derive from a sub-class (as Rupert and you have both pointed out, they have functionality in common relating to sending, receiving, merging, and colouring), which would have some function like !GetColour(char[3]). Code in the level above can decide how to combine and serialise the colours.
The glyphs are sadly the fly in the oitment, as they result in the darkening of a non-glyph.

**Comment by schmie** [comment]
**Comment by hywel on 23 Sep 2011 14:10 UTC**
Replying to [nick](comment:5):
> Replying to [hywel](comment:4):
> > That bitfield limits us to a certain size of image. I can imagine situations where we'd want to go beyond that. Also, the streakline data includes 2 floats and an int. The current implementation reads, to me, like a computer scientist fiddling with his bits, if you'll forgive the expression. If you haven't read it before, I highly recommend a Knuth paper called "Literate Programming" which certainly convinces me that the new design is better than the old.
> Fair point - I'd forgotten about the extra fields for the streakline data, which probably does make the changes much more advantageous. I'm very much in favour of coding to make things look nicer. I had just assumed that our changes weren't allowed to reduce the performance of hemeLB without good reason.

Small performance reduction for better engineered code is a good reason. And messages have to be significantly larger before you'll notice the performance difference.

> > If we're only displaying streaklines on the client, why should we render, send and compose this other stuff at all? I mean this rhetorically - obviously we shouldn't be sending it. Remember too that the diff-test is not the use-case we're aiming for, it's a regression test for the research (i.e. non-steered) case.
> I'd assumed that the streaklines would not be rendered without the ray tracing data, as it'd be hard to infer anything without the vessels in the background.

Probably, at the moment.

> > Nick, it's possible to have only the class that creates and consumes the data know about the type and nothing intermediate. The templating can therefore end at the boundary to the class. Run-type polymorphism is one way to handle this, encapsulation of the datatype is another (and the one I prefer).
> How the latter would work?

You'd hand it a pointer and tell it to do what it needs to send data, then on the other side hand it a pointer and tell it to do what it needs to receive data from the pointer. External stuff only needs to know where its data starts and ends, nothing about what's inside.

> > Also note that our designs should not limit themselves to consider rays, glyphs and streaklines. For instance consider how much copy-and-pasting would be involved to have both ray-tracers running in parallel, visualising the results on the steering client (if the answer is > 0 copy-and-pasting then we're doing something wrong). Adding a new imaging technique should be as easy as writing a self-contained class, registering it in one place then running the simulation.
> Except that the data into the class varies depending on the type of imaging.

It shouldn't do, you should be able to give it a !LatticeData object and whatever else and tell the class to take what it needs. If you really wanted to avoid dynamic polymorphism, you could do this using the CRTP to get statically-typed objects that obey an interface.

> > We haven't investigated parallel non-blocking communications. It's possible it would make the code faster, it's also possible that it wouldn't (depending on the MPI implementation - each communication has a constant latency time of about 10 micro s; if it takes 9 micro s to do computation before the next parallel non-blocking communication, we've saved nothing). The central aim of this work is to improve the engineering of the visualisation code. 
> Fair enough
> 
> > The original design is Marco's, rather than Rupert's. Something similar might be used for the actual communication, but there would be a conversion step beforehand (this is what I was talking about on comment:6:ticket:7). These would be converted back on the other side before being merged. Each pixel would be responsible for merging itself.
> Fair enough - this is not far off what's already happening with the ray data, except that the raydata from all the rays are prematurely added to ColPixel.

Indeed, it's not a massive change from a computational point of view, but it's a significant one from a structural perspective.

> > You're right that the !ColPixels contain the code for rendering themselves (as well as serializing themselves) and that I hadn't described that here. My intention was to have each pixeltype derive from a sub-class (as Rupert and you have both pointed out, they have functionality in common relating to sending, receiving, merging, and colouring), which would have some function like !GetColour(char[3]). Code in the level above can decide how to combine and serialise the colours.
> The glyphs are sadly the fly in the oitment, as they result in the darkening of a non-glyph.

Only at the moment - the interaction between visualisation types on the level of the pixel is a bad smell. Like I say, code in the level above can decide how to combine colours, and that includes interactions.

**Comment by schmie** [comment]
**Comment by nick on 23 Sep 2011 14:26 UTC**
Replying to [hywel](comment:6):
> Small performance reduction for better engineered code is a good reason. And messages have to be significantly larger before you'll notice the performance difference.
While I'll would full agree, the logical conclusion would be to use a language with garbage collection and reflection. I also understood the communication was the rate limiting step.
 
> > > Nick, it's possible to have only the class that creates and consumes the data know about the type and nothing intermediate. The templating can therefore end at the boundary to the class. Run-type polymorphism is one way to handle this, encapsulation of the datatype is another (and the one I prefer).
> > How the latter would work?
> 
> You'd hand it a pointer and tell it to do what it needs to send data, then on the other side hand it a pointer and tell it to do what it needs to receive data from the pointer. External stuff only needs to know where its data starts and ends, nothing about what's inside.
I thought the point of using MPI and why we used user defined data type rather than sizeof() was that it was architecture independent.
 
 
> It shouldn't do, you should be able to give it a !LatticeData object and whatever else and tell the class to take what it needs. If you really wanted to avoid dynamic polymorphism, you could do this using the CRTP to get statically-typed objects that obey an interface.
Fair enough - I think the ray tracer is a special case in that it has the UpdateClusterVoxel method.

> Only at the moment - the interaction between visualisation types on the level of the pixel is a bad smell. Like I say, code in the level above can decide how to combine colours, and that includes interactions.
What about using the decorator pattern? You could also add extra classes that allow for transparency, eg having the streaklines at 50% opacity?

**Comment by schmie** [comment]
**Comment by hywel on 26 Sep 2011 07:12 UTC**
Replying to [nick](comment:7):
> Replying to [hywel](comment:6):
> > Small performance reduction for better engineered code is a good reason. And messages have to be significantly larger before you'll notice the performance difference.
> While I'll would full agree, the logical conclusion would be to use a language with garbage collection and reflection. I also understood the communication was the rate limiting step.

Communication is often rate-limiting; I've done some work recently that reduces that. Making a message bigger has comparatively little effect. Using a garbage-collected language would give a big reduction in speed; it's possible to have perfectly well-engineered code with C++. 

> > > > Nick, it's possible to have only the class that creates and consumes the data know about the type and nothing intermediate. The templating can therefore end at the boundary to the class. Run-type polymorphism is one way to handle this, encapsulation of the datatype is another (and the one I prefer).
> > > How the latter would work?
> > You'd hand it a pointer and tell it to do what it needs to send data, then on the other side hand it a pointer and tell it to do what it needs to receive data from the pointer. External stuff only needs to know where its data starts and ends, nothing about what's inside.
> I thought the point of using MPI and why we used user defined data type rather than sizeof() was that it was architecture independent.

There are many reasons for using MPI -- architecture-independence is one. And we would still use a user-defined datatype. The C++ type for this is an MPI_Datatype - the fact that it's actually 3 ints and a float or whatever is fully internal to the user-defined type (and all other bits of code can just take your MPI_Datatype and don't need to know what it defines).

>  
> > It shouldn't do, you should be able to give it a !LatticeData object and whatever else and tell the class to take what it needs. If you really wanted to avoid dynamic polymorphism, you could do this using the CRTP to get statically-typed objects that obey an interface.
> Fair enough - I think the ray tracer is a special case in that it has the UpdateClusterVoxel method.
> 
> > Only at the moment - the interaction between visualisation types on the level of the pixel is a bad smell. Like I say, code in the level above can decide how to combine colours, and that includes interactions.
> What about using the decorator pattern? You could also add extra classes that allow for transparency, eg having the streaklines at 50% opacity?

That's certainly one plan. It's not something we need at the moment but it might well be how we do it in the future.

**Comment by schmie** [comment]
**Comment by hywel on 7 Oct 2011 14:38 UTC**
Diff attached. Unit-tests still pass, diff-test produces the same images (modulo the streakline bug described on ticket:4). The memory usage when running the diffTest used to be around 500MB for me, now it's around 25MB. The price of this is an extra O(Log(num. pixels)) lookup whenever a pixel is accessed.

Notes:
 * I'm pretty sure this diff breaks the case when stress-type is set to Von Mises (rather than Shear). The desired behaviour on the client is unknown for this case, so I'm suggesting we work that out at a later date and fix any problems created then.
 * Much of this patch centres around the idea of a !PixelSet, a templated object with a vector of pixels and a map of the co-ordinates of a pixel to the index in the vector. It also handles combining.
 * We also have !PixelSetStores, which exist to prevent us dynamically allocating and deallocating !PixelSets all the time. I ran the diffTest creating an image each iteration and the memory usage didn't go over 28MB (suggesting we're not leaking with each image created).
 * I got rid of some commented-out code.
 * I think I've made it easier to add new visualisers, but it's not yet as simple as just adding your new drawer to a list. Having spoken to Rupert, I'm not doing this now because it's not a priority.
 * Each pixel type is MPI-communicated and stored separately to make it easier to plug and play these.
 * Ignore !DrawablePixel - that wasn't meant to be in the diff and I've now deleted the file.
 * The Screen object now represents a screen only. Since making this diff I've moved the declaration of COLOURED_PIXELS_MAX to Screen from !ImageSendComponent, and we now make sure that the screen isn't resized to have more pixels than this limit. I can upload the diff for this if required.

Assigning to Miguel to review the code.

**Comment by schmie** [comment]
**Comment by miguel on 10 Oct 2011 16:16 UTC**
Good stuff overall, just a few minor points (number ranges refer to diff file lines):
 * 12-15: don't seem to be well indented.
 * 149-150: it is probably self-explanatory, but adding a description after @param param_name makes doxygen happier
 * 237: Does constant 4 mean sizeof(int)? I would suggest not having datatypes sizes hardcoded for portability. Also, unless we allow maxSendSize to take negative numbers, it should probably be an unsigned.
 * Good use of `const` here and there.
 * When dereferencing iterators there's no agreement between `(*iter).second` and `iter->second` in `Control.cc`
 * 1786-1792 same comment about doxygen as above.
 * 1803 Is it possible to have more meaningful names for `int d = incNE;`?
 * Good naming in class `Rendering`
 * Tabs in `/Code/vis/Rules.mk`

**Comment by schmie** [comment]
**Comment by hywel on 13 Oct 2011 11:24 UTC**
Replying to [miguel](comment:10):
> Good stuff overall, just a few minor points (number ranges refer to diff file lines):
>  * 12-15: don't seem to be well indented.
Done
>  * 149-150: it is probably self-explanatory, but adding a description after @param param_name makes doxygen happier
Done
>  * 237: Does constant 4 mean sizeof(int)? I would suggest not having datatypes sizes hardcoded for portability. Also, unless we allow maxSendSize to take negative numbers, it should probably be an unsigned.
4 is the size of an XDR int, so is more portable than sizeof(int) would be here. I added a new static const ```XdrIntLength``` and referred to that instead, so it's clearer.
>  * Good use of `const` here and there.
>  * When dereferencing iterators there's no agreement between `(*iter).second` and `iter->second` in `Control.cc`
Yup, sorry. I've gone for (*iter).property throughout as I think it's that little bit clearer.
>  * 1786-1792 same comment about doxygen as above.
Yup, done.
>  * 1803 Is it possible to have more meaningful names for `int d = incNE;`?
I've added some comments to the calling code and the helper function. It took a while, but I think I now understand it and have renamed variables and added comments in-line with what I think.
>  * Good naming in class `Rendering`
>  * Tabs in `/Code/vis/Rules.mk`
Those were corrections to the original tabs, so they're actually correct.

**Comment by schmie** [comment]
**Modified by hywel on 13 Oct 2011 11:38 UTC**

**Comment by schmie** [comment]
**Comment by hywel on 13 Oct 2011 15:23 UTC**
Pushed to Entropy as changeset 003309bde8c9.
